### PR TITLE
add SMAPE objective function

### LINKED ISSUE #624: can add a mape objective function in lightgbm?
Such as the title.
reference to：https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

could show the mape objective function code segment in python now.

### PR DESCRIPTION
**Pull request by lakshayg**
closes #624

### CODE REVIEW

File: src/metric/regression_metric.hpp [hunk 1]
```diff
@@ -171,6 +171,19 @@ class L1Metric: public RegressionMetric<L1Metric> {
   }
 };
 
+/*! \brief MAPE loss for regression task */
+class MAPEMetric: public RegressionMetric<MAPEMetric> {
+public:
+  explicit MAPEMetric(const MetricConfig& config) :RegressionMetric<MAPEMetric>(config) {}
+
+  inline static double LossOnPoint(label_t label, double score, const MetricConfig&) {
+    return 100 * std::fabs(score - label) / label;
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: label could be zero.

File: src/objective/regression_objective.hpp [hunk 2]
```diff
@@ -161,6 +161,55 @@ class RegressionL1loss: public RegressionL2loss {
...
+    if (weights_ == nullptr) {
+      #pragma omp parallel for schedule(static)
+      for (data_size_t i = 0; i < num_data_; ++i) {
+        const double diff = score[i] - label_[i];
+        if (diff >= 0.0f) {
+          gradients[i] = 1.0f;
+        } else {
+          gradients[i] = -1.0f;
+        }
+        hessians[i] = static_cast<score_t>(0.0);
```
Comment (guolinke) [inline src/objective/regression_objective.hpp]: cannot use zero hessians, since the output is grad/hess.

File: src/metric/regression_metric.hpp (hunk 1 above)
Comment (lakshayg) [reply src/metric/regression_metric.hpp]: What do you suggest in such a case?

Comment (guolinke) [reply src/metric/regression_metric.hpp]: maybe, when label=0 and score=0, return 0, 
otherwise return score directly.

Comment (guolinke) [reply src/metric/regression_metric.hpp]: @Laurae2 any suggestions  ?

Comment (lakshayg) [reply src/metric/regression_metric.hpp]: Do you mean

```
if (label == 0 && score == 0) => return 0;
if (label == 0 && score != 0) => return score;
otherwise return (score - label) / label;
```

File: src/objective/regression_objective.hpp (hunk 2 above)
Comment (lakshayg) [reply src/objective/regression_objective.hpp]: What should be done in this case?

Comment (guolinke) [reply src/objective/regression_objective.hpp]: you can use some approximate loss function. e.g. 1. https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error . 
2. use L2 form: loss = (y-label)^2/label .
BTW, your gradient and hess actually is the L1, not the MAPE.

Comment (lakshayg) [reply src/objective/regression_objective.hpp]: > BTW, your gradient and hess actually is the L1, not the MAPE.
Yes, I just realized that I messed up. Will update shortly

**Comment by guolinke** [comment]
I think it is an interesting problem for the loss with hessian = 0.
for example, the L1, quantile, mape and huber. 
Are there some working solutions ? 
ping @Laurae2 @henry0312 .

**Comment by lakshayg** [comment]
To prevent dealing with label == 0 or hessian == 0, I have implemented [SMAPE][1] instead of [MAPE][2].

[1]: https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error
[2]: https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

**Comment by lakshayg** [comment]
The CI error is due to SSLError. Can one of the devs have a look, it seems unrelated to this change.

File: src/objective/regression_objective.hpp [hunk 3]
```diff
@@ -161,6 +161,73 @@ class RegressionL1loss: public RegressionL2loss {
...
+        if (diff == 0) {
+          gradients[i] = 0.0f;
+          hessians[i] = 2.0f * weights_[i] / sumabs;
+        } else {
+          const double absdiff = std::fabs(diff);
+          const double sumabs2 = sumabs * sumabs;
+          const double sumabs3 = sumabs2 * sumabs;
+          const double t1 = (2.0f / sumabs3 - (2.0f * (score[i] == 0)) / (label_[i] * label_[i]));
+          const double t2 = 2.0f  * sgn(label_[i]) * sgn(diff) / sumabs2;
+          gradients[i] = ((sgn(diff) * sumabs - sgn(score[i]) * absdiff)) * weights_[i] / sumabs2;
```
Comment (guolinke) [inline src/objective/regression_objective.hpp]: you can use `std::copysign` .

Comment (lakshayg) [reply src/objective/regression_objective.hpp]:
```
sgn(a) * b != std::copysign(b, a) // when a = 0
```

Should I still replace sgn with copysign? Perhaps there is a better place to keep copysign so that it can be used in other places as well. A number of places in code can make use of this function rather than having branches.

Comment (guolinke) [reply src/objective/regression_objective.hpp]: copysign is more efficient, it will avoid the if...else check.

Comment (lakshayg) [reply src/objective/regression_objective.hpp]: I meant that the function I wrote relies on the fact that `sgn(0) = 0`. This is not the case with copysign. I am also unsure about how copysign will avoid if-else. Can you please elaborate?

Comment (guolinke) [reply src/objective/regression_objective.hpp]: you can see the compiled Assemble code, copysign will use the cpu instructions to get sign directly

Comment (guolinke) [reply src/objective/regression_objective.hpp]: i think copysign is safe for 0 . you can test it

Comment (lakshayg) [reply src/objective/regression_objective.hpp]:
```
  std::cout << std::copysign(1, 0) << std::endl;
```

The above statement print 1 and not 0

Comment (lakshayg) [reply src/objective/regression_objective.hpp]: As far as branching in `sgn` is concerned, the version I wrote does not have any branches https://godbolt.org/g/AUnpfi

**Comment by Laurae2** [comment]
@guolinke @lakshayg Gradient and Hessian of SMAPE:

```r
grad <- function(x, y) {
  (2 * y * (x - y)) / ((x + y) * (x + y) * abs(x - y))
}

hess <- function(x, y) {
  (-4 * y * (x - y)) / ((x + y) * (x + y) * (x + y) * abs(x - y))
}
```

Tested and working in [Kaggle Web Traffic Competition](https://www.kaggle.com/c/web-traffic-time-series-forecasting/discussion), identical results to the symbolic R math function conversion.

Below is the symbolic math function conversion from R:

```r
grad <- function(x, y) {
  .e2 <- abs(x) + y
  .e3 <- x - y
  (sign(.e3) - abs(.e3) * sign(x)/.e2)/.e2
}

hess <- function(x, y) {
  .e2 <- abs(x) + y
  .e3 <- sign(x)
  .e4 <- x - y
  -(2 * ((sign(.e4) - abs(.e4) * .e3/.e2) * .e3/.e2^2))
}
```

To guarantee a converging Gradient/Hessian of MAPE (because of hessian = 0), it requires to approximate a function which is both similar to the MAPE (linearly) and a quadratic version of itself (such as how Fair loss, etc. are done for L1/MAE loss). Otherwise, there is no guarantee MAPE can converge (it will explode in most of cases). SMAPE usually is not an alternative as MAPE has the very strong intrinsic feature of underpredicting.

When label = 0, loss should be considered 0 otherwise it explodes to infinity. When loss is 0, the associated gradients and hessians should be also 0.

**Comment by lakshayg** [comment]
@Laurae2 Thanks a lot for the detailed comment. Here is what I think should be the way to go then:

1) Update the SMAPE grad and hessian with what you suggested
2) This PR should not  close the issue since you said that SMAPE is not an alternative to MAPE (could you suggest an alternative?)
3) Close the issue once an alternative to MAPE is implemented (implementing MAPE itself is probably not a good idea since you say that it mostly explodes)

What are your views?

File: src/objective/regression_objective.hpp (hunk 3 above)
Comment (guolinke) [reply src/objective/regression_objective.hpp]: okay, great

**Comment by lakshayg** [comment]
@Laurae2 Could you clarify what `x` and `y` mean? Is `x = prediction` and `y = label`?

**Comment by Laurae2** [comment]
@lakshayg It must be `x = prediction` and `y = label`.

Test:

```r
library(xgboost)

data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')

dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label + 1)
dtest <- xgb.DMatrix(agaricus.test$data, label = agaricus.test$label + 1)
watchlist <- list(train = dtrain, eval = dtest)

param <- list(max_depth = 2, eta = 1, silent = 1, nthread = 2, 
              objective = "binary:logistic", eval_metric = "auc")
bst <- xgb.train(param, dtrain, nrounds = 2, watchlist)

grad <- function(x, y) {
  (2 * y * (x - y)) / ((x + y) * (x + y) * abs(x - y))
}

hess <- function(x, y) {
  (-4 * y * (x - y)) / ((x + y) * (x + y) * (x + y) * abs(x - y))
}

smape_obj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  grad <- grad(labels, preds)
  hess <- hess(labels, preds)
  return(list(grad = grad, hess = hess))
}

smape_loss <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  err <- mean((abs(preds - labels)) / ((abs(preds) + abs(labels)) / 2))
  return(list(metric = "smape", value = err))
}

# These functions could be used by passing them either:
#  as 'objective' and 'eval_metric' parameters in the params list:
param <- list(max_depth = 2, eta = 0.1, nthread = 1, 
              objective = smape_obj, eval_metric = smape_loss)
bst <- xgb.train(param, dtrain, nrounds = 10, watchlist)
```

Yields convergence:

```r
[1]	train-smape:0.786086	eval-smape:0.786127 
[2]	train-smape:0.654798	eval-smape:0.655078 
[3]	train-smape:0.529639	eval-smape:0.530132 
[4]	train-smape:0.410331	eval-smape:0.411013 
[5]	train-smape:0.296609	eval-smape:0.297458 
[6]	train-smape:0.216386	eval-smape:0.216352 
[7]	train-smape:0.237770	eval-smape:0.237861 
[8]	train-smape:0.254103	eval-smape:0.254304 
[9]	train-smape:0.266287	eval-smape:0.266585 
[10]	train-smape:0.298564	eval-smape:0.298592 
```

Inverting `preds` and `labels` in `smape_obj` won't converge:

```r
[1]	train-smape:0.923793	eval-smape:0.923567 
[2]	train-smape:0.923793	eval-smape:0.923567 
[3]	train-smape:0.923793	eval-smape:0.923567 
[4]	train-smape:0.923793	eval-smape:0.923567 
[5]	train-smape:0.923793	eval-smape:0.923567 
[6]	train-smape:0.923793	eval-smape:0.923567 
[7]	train-smape:0.923793	eval-smape:0.923567 
[8]	train-smape:0.923793	eval-smape:0.923567 
[9]	train-smape:0.923793	eval-smape:0.923567 
[10]	train-smape:0.923793	eval-smape:0.923567
```

(due to the labels, using SMAPE is way too much inappropriate but this is just for the example - using a simple quadratic loss yields 0.01 SMAPE nearly instantly)

**Comment by lakshayg** [comment]
@Laurae2 Can you explain how you came up with this derivative? I tried using wolfram alpha but could not reduce its output to the gradient that you specified. Are there any assumptions that I might be missing?

http://www.wolframalpha.com/input/?i=differentiate+2*+%7Cx-y%7C+%2F+(%7Cx%7C+%2B+%7Cy%7C)

**Comment by Laurae2** [comment]
@lakshayg use this: https://www.derivative-calculator.net/

Input this formula: `(abs(x - y)) / ((abs(x) + abs(y)) / 2)`

Request the second derivative:

image

Interesting plot explaining why it's very hard for SMAPE with hessian (change it to 2 and you will see why):

image

**Comment by guolinke** [comment]
I think we can continue the discussion here: https://github.com/Microsoft/LightGBM/issues/979

File: src/metric/regression_metric.hpp [hunk 4]
```diff
@@ -171,6 +171,19 @@ class L1Metric: public RegressionMetric<L1Metric> {
   }
 };
 
+/*! \brief SMAPE loss for regression task */
+class SMAPEMetric: public RegressionMetric<SMAPEMetric> {
+public:
+  explicit SMAPEMetric(const MetricConfig& config) :RegressionMetric<SMAPEMetric>(config) {}
+
+  inline static double LossOnPoint(label_t label, double score, const MetricConfig&) {
+    return std::fabs(score - label) / (std::fabs(label) + std::fabs(score));
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: I think ```std::fabs(label) + std::fabs(score)``` also could be zero.

**Comment by guolinke** [comment]
@Laurae2 Is grad/hess of current implementation okay ?

**Comment by lakshayg** [comment]
I think std::fabs(label) + std::fabs(score) also could be zero.

Yes you are correct. I am working on implementing this again and will update the PR sometime this weekend.

File: src/metric/regression_metric.hpp (hunk 4 above)
Comment (lakshayg) [reply src/metric/regression_metric.hpp]: Yes you are correct. I am working on implementing this again and will update the PR sometime this weekend.

Comment (guolinke) [reply src/metric/regression_metric.hpp]: great! thanks very much

**Comment by Laurae2** [comment]
@lakshayg Which exact formula are you using for grad/hess?

I'd like to test them as custom losses to test in xgboost/LightGBM to check.

@guolinke It seems to be OK when looking at them, didn't check in depth though.

**Comment by guolinke** [comment]
@lakshayg please update the document accordingly.

**Comment by Laurae2** [comment]
@lakshayg any news?

@guolinke I think SMAPE should use the gradient only (since you managed to make hessian less LightGBM). With hessian, it is unstable.

**Comment by lakshayg** [comment]
Currently busy with work, I guess I will be able to work this/next weekend. Sorry for the delay.

**Comment by StrikerRUS** [comment]
@lakshayg Are you still interested in this?

**Comment by lakshayg** [comment]
Sorry for the delay in responding. Yes, I'm still interested. Will give this another try this weekend.

On 31 May 2018 12:22 pm, Nikita Titov <notifications@github.com> wrote:

@lakshayg<https://github.com/lakshayg> Are you still interested in this?

**Comment by lakshayg** [comment]
Sorry, I have not been able to find time to continue working on this. I will close the PR for now and will send in a different PR if/when I am able to get back to it.
